### ISSUE TITLE
Serious issue in cookie max-age generation in Play 2.3.9

### ISSUE THREAD

**Issue by alexdupre** [playframework/playframework#4432]
With new version of play/netty (2.3.9/3.9.8) the generation of cookie max-age field is borked.

A simple controller like:
```
  def test = Action { request =>
    Ok.withCookies(Cookie("test", "look at maxAge", Some(2592000)))
  }
```

generates a response with a negative max-age:

```
HTTP/1.1 200 OK
Content-Length: 0
Set-Cookie: test=look at maxAge; Max-Age=-1702966; Expires=Tue, 21 Apr 2015 20:57:00 GMT; Path=/; HTTPOnly
```

**Comment by jroper** [comment]
Ok, so the initial code that generates the header:

```
scala> import play.api.mvc._
import play.api.mvc._

scala> import Results._
import Results._

scala> Ok.withCookies(Cookie("test", "look at maxAge", Some(2592000)))
res0: play.api.mvc.Result = Result(200, Map(Set-Cookie -> test=look at maxAge; Max-Age=2592000; Expires=Wed, 22 Apr 2015 15:23:34 GMT; Path=/; HTTPOnly))
```

`Max-Age` is correct, `Expires` is not.  This may be a bug in Netty.  Play actually later on decodes this using Netty again in `NettyResultStreamer`, which probably decodes the `Expires` and uses that in preference to `Max-Age`.  We can trigger Play's reparsing using `withCookies` again:

```
scala> Ok.withCookies(Cookie("test", "look at maxAge", Some(2592000))).withCookies()
res4: play.api.mvc.Result = Result(200, Map(Set-Cookie -> test=look at maxAge; Max-Age=-1702966; Expires=Wed, 22 Apr 2015 15:28:20 GMT; Path=/; HTTPOnly))
```

And a third time, it's now stable:

```
scala> Ok.withCookies(Cookie("test", "look at maxAge", Some(2592000))).withCookies().withCookies()
res5: play.api.mvc.Result = Result(200, Map(Set-Cookie -> test=look at maxAge; Max-Age=-1702966; Expires=Wed, 22 Apr 2015 15:28:51 GMT; Path=/; HTTPOnly))
```

So it's not the reparsing, but the initial generation of `Expires` that's the problem.

**Comment by jroper** [comment]
Ok, I've found the bug, it only happens for values greater than `Int.MaxInt / 1000`, ie values greater than 24 days.

**Comment by alexdupre** [comment]
Yeah, this is the netty bug. The following line:
```
Date expires = new Date(cookie.maxAge() * 1000 + System.currentTimeMillis());
```
should be:
```
Date expires = new Date(cookie.maxAge() * 1000L + System.currentTimeMillis());
```

**Comment by jroper** [comment]
Fix for Netty 3.9 branch: https://github.com/netty/netty/pull/3777

**Comment by jroper** [comment]
A work around for this is to downgrade to Netty 3.9.6.  You won't get the cookies security vulnerability fix, however, that's only an issue in very specific circumstances, for example, if using Play's CSRF protection **and** setting **csrf.cookie.name** so that it uses cookies instead of the session to store CSRF tokens.

**Comment by jroper** [comment]
ETA for Netty release is looking like tomorrow.

**Comment by tsuyoshizawa** [comment]
This problem has been fixed at #4454, but only master branch yet.
Do you forget to take in the commit to 2.3.x branch?

Some users in Play 2.3 would hope to be fixed this problem soon.

**Comment by jroper** [comment]
Hi @tsuyoshizawa, in master, we actually have the code copied out of netty. Play 2.3.x doesn't, it uses netty. So we can't backport that change to 2.3.x, we can only wait for netty to cut a new release.

**Comment by tsuyoshizawa** [comment]
@jroper, that was my misunderstanding.

I wait the new release by netty.
Thanks!

**Comment by mkurz** [comment]
Netty 3.9.9 has finally been released. See #4929
I think we could now release Play 2.3.10

**Comment by marcospereira** [comment]
Fixed by #4929.
